x-pack/apm-server/sampling/eventstorage/storage_manager.go

```diff
@@ -53,288 +75,188 @@ type StorageManager struct {
+func (sm *StorageManager) savePartitionID(pid int) error {
+	return sm.decisionDB.Set([]byte(partitionerMetaKey), []byte(fmt.Sprintf(`{"id":%d}`, pid)), pebble.NoSync)
 }
 
-// Size returns the db size
-func (s *StorageManager) Size() (lsm, vlog int64) {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
-	return s.db.Size()
+func (sm *StorageManager) Size() (lsm, vlog int64) { // FIXME: stop calling it vlog
+	// FIXME: separate WAL usage?
+	return int64(sm.eventDB.Metrics().DiskSpaceUsage() + sm.decisionDB.Metrics().DiskSpaceUsage()), 0
 }
 
-func (s *StorageManager) NewTransaction(update bool) *badger.Txn {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
-	return s.db.NewTransaction(update)
+func (sm *StorageManager) DiskUsage() uint64 {
+	return sm.eventDB.Metrics().DiskSpaceUsage() + sm.decisionDB.Metrics().DiskSpaceUsage()
+}
+
+func (sm *StorageManager) StorageLimit() uint64 {
+	return sm.storageLimit.Load()
+}
+
+func (sm *StorageManager) Flush() error {
+	return errors.Join(sm.eventDB.Flush(), sm.decisionDB.Flush())
+}
+
+func (sm *StorageManager) Close() error {
+	return sm.close()
+}
+
+func (sm *StorageManager) close() error {
+	return errors.Join(sm.eventDB.Close(), sm.decisionDB.Close())
+	// FIXME: there seems to be an error with pebble Flush hanging indefinitely
+	//return errors.Join(sm.eventDB.Flush(), sm.decisionDB.Flush(), sm.eventDB.Close(), sm.decisionDB.Close())
+}
+
+// Reload flushes out pending disk writes to disk by reloading the database.
+// For testing only.
+func (sm *StorageManager) Reload() error {
+	if err := sm.close(); err != nil {
+		return err
+	}
+	return sm.reset()
 }
 
 // Run has the same lifecycle as the TBS processor as opposed to StorageManager to facilitate EA hot reload.
-func (s *StorageManager) Run(stopping <-chan struct{}, gcInterval time.Duration, ttl time.Duration, storageLimit uint64, storageLimitThreshold float64) error {
+func (sm *StorageManager) Run(stopping <-chan struct{}, ttl time.Duration, storageLimit uint64) error {
 	select {
 	case <-stopping:
 		return nil
-	case s.runCh <- struct{}{}:
+	case sm.runCh <- struct{}{}:
 	}
 	defer func() {
-		<-s.runCh
+		<-sm.runCh
 	}()
 
-	g := errgroup.Group{}
-	g.Go(func() error {
-		return s.runGCLoop(stopping, gcInterval)
-	})
-	g.Go(func() error {
-		return s.runDropLoop(stopping, ttl, storageLimit, storageLimitThreshold)
-	})
-	return g.Wait()
+	sm.storageLimit.Store(storageLimit)
+
+	return sm.runTTLGCLoop(stopping, ttl)
 }
 
-// runGCLoop runs a loop that calls badger DB RunValueLogGC every gcInterval.
-func (s *StorageManager) runGCLoop(stopping <-chan struct{}, gcInterval time.Duration) error {
-	// This goroutine is responsible for periodically garbage
-	// collecting the Badger value log, using the recommended
-	// discard ratio of 0.5.
-	ticker := time.NewTicker(gcInterval)
+func (sm *StorageManager) runTTLGCLoop(stopping <-chan struct{}, ttl time.Duration) error {
+	ttlGCInterval := ttl / partitionsPerTTL
```

Comment:
`partitionsPerTTL` gets increased by 1 active and 1 inactive partition; e.g. if `partitionsPerTTL==1`, overall there are 2 active and 1 inactive partition. Wouldn't it make sense to divide by number of active partitions here as there is 1 partition buffer already from the inactive one?

Reply:
I've attached the image of how it currently works in the PR description, but I'll attach a copy here too. 
![image](https://github.com/user-attachments/assets/388b35ab-b346-4f16-b039-80a0e3acfe17)

Let's imagine we adopt your suggestion that ttlGCInterval is `ttl / (partitionsPerTTL+1)` instead of `ttl / partitionsPerTTL`, then each partition covers 10m instead of 15m, then the 3 active partitions only cover 30m backwards from the ending time of the current partition. That also means if we are only 5mins into this current partition, we're only keeping 25mins of events while we SHOULD be keeping TTL=30m of events. It is not correct.

Reply:
You are right; with your approach we are erroring on the side of keeping the data up to one ttl slice longer, which makes sense!